lib/rubocop/rspec/top_level_group.rb

```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module RSpec
+    # Helper methods for top level example group cops
+    module TopLevelGroup
+      extend NodePattern::Macros
+
+      def_node_matcher :example_or_shared_group?,
+        (Language::ExampleGroups::ALL + Language::SharedGroups::ALL).block_pattern
+
+      def on_block(node)
+        return unless respond_to?(:on_top_level_group)
+        return unless top_level_group?(node)
+
+        on_top_level_group(node)
+      end
+
+      private
+
+      def top_level_group?(node)
+        top_level_groups.include?(node)
+      end
+
+      def top_level_groups
+        @top_level_groups ||= top_level_nodes.select { |n| example_or_shared_group?(n) }
```

Comment:
Can't say off the top of my head if there is one single instance of a cop for all the files being checked, or a new cop instance is created for every file.
Wondering if this might change suddenly in the future as a flyweight optimization and cops using this module will only take action on the first file they analyze.
This might be checked by hand, but I'm not sure if this can be covered by our specs.

Reply:
The only precautionary measure I can think of is a check that `processed_source` or its `ast` is still the same object.
I hope there are no plans to reuse those objects between files :D

Reply:
> Can't say off the top of my head if there is one single instance of a cop for all the files being checked, or a new cop instance is created for every file.

Right now a new cop instance is created for every file

Reply:
Cool, please disregard then. 👍 
There's no need for us to be defensive.

Reply:
> Wondering if this might change suddenly in the future as a flyweight optimization and cops using this module will only take action on the first file they analyze.

It's definitely an issue.

I noticed at least one module (`SurroundingSpace`) which memoizes some values in cop's instance variables so it's an already existing issue :-)

https://github.com/rubocop-hq/rubocop/blob/master/lib/rubocop/cop/mixin/surrounding_space.rb#L47-L56